rust/cubesql/cubesql/src/sql/session_manager.rs

```diff
@@ -41,60 +50,72 @@ impl SessionManager {
         protocol: DatabaseProtocol,
         client_addr: String,
         client_port: u16,
-    ) -> Arc<Session> {
+        extra_id: Option<String>,
+    ) -> Result<Arc<Session>, CubeError> {
         let connection_id = self.last_id.fetch_add(1, Ordering::SeqCst);
 
-        let sess = Session {
+        let session_ref = Arc::new(Session {
             session_manager: self.clone(),
             server: self.server.clone(),
             state: Arc::new(SessionState::new(
                 connection_id,
+                extra_id.clone(),
                 client_addr,
                 client_port,
                 protocol,
                 None,
                 Duration::from_secs(self.server.config_obj.auth_expire_secs()),
                 Arc::downgrade(self),
             )),
-        };
-
-        let session_ref = Arc::new(sess);
+        });
 
         let mut guard = self.sessions.write().await;
 
-        guard.insert(connection_id, session_ref.clone());
+        if let Some(extra_id) = extra_id {
+            if guard.uid_to_session.contains_key(&extra_id) {
+                return Err(CubeError::user(format!(
+                    "Session cannot be created, because extra_id: {} already exists",
+                    extra_id
+                )));
+            }
 
-        session_ref
-    }
+            guard.uid_to_session.insert(extra_id, session_ref.clone());
+        }
 
-    pub async fn stat_activity(self: &Arc<Self>) -> Vec<SessionStatActivity> {
-        let guard = self.sessions.read().await;
+        guard.sessions.insert(connection_id, session_ref.clone());
 
-        guard
-            .values()
-            .map(Session::to_stat_activity)
-            .collect::<Vec<SessionStatActivity>>()
+        Ok(session_ref)
     }
 
-    pub async fn process_list(self: &Arc<Self>) -> Vec<SessionProcessList> {
+    pub async fn map_sessions<T: for<'a> From<&'a Arc<Session>>>(self: &Arc<Self>) -> Vec<T> {
```

Comment:
Current `map_sessions` usages don't need `Arc` part, `From<&'a Session>` should be enough.
`Arc` here can be useful (e.g. to be able to clone `Arc` instance), but as it is now it's bit confusing to me